### PR TITLE
remove `type coercion` for physical ScalarFunction

### LINKED ISSUE #3731: support `type coercion` for ScalarFunction expr in the logical phase
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
 part of https://github.com/apache/arrow-datafusion/issues/3582

move the type coercion for the expr: ScalarFunction from the physical phase to the logical phase.
**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by liukun4515**
# Which issue does this PR close?

Closes #3731 

 # Rationale for this change

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/physical-expr/src/functions.rs [hunk 1]
```diff
@@ -58,23 +58,20 @@ pub fn create_physical_expr(
     input_schema: &Schema,
     execution_props: &ExecutionProps,
 ) -> Result<Arc<dyn PhysicalExpr>> {
-    let coerced_phy_exprs =
```
Comment (liukun4515) [inline datafusion/physical-expr/src/functions.rs:61]: remove type coercion in the physical phase

File: datafusion/optimizer/src/type_coercion.rs [hunk 2]
```diff
@@ -310,18 +310,6 @@ impl ExprRewriter for TypeCoercionRewriter {
                 };
                 Ok(expr)
             }
-            Expr::ScalarUDF { fun, args } => {
```
Comment (liukun4515) [inline datafusion/optimizer/src/type_coercion.rs:313]: just move these code

**Comment by alamb** [comment]
The CI checks appear to be failing

**Comment by liukun4515** [comment]
> The CI checks appear to be failing

Some errors about import in the test, I have fixed them

**Review by andygrove — COMMENTED** [review]
LGTM. Thanks @liukun4515

File: datafusion/physical-expr/src/functions.rs [hunk 3]
```diff
@@ -2872,4 +2874,17 @@ mod tests {
 
         Ok(())
     }
+
+    // Helper function
+    // The type coercion will be done in the logical phase, should do the type coercion for the test
```
Comment (alamb) [inline datafusion/physical-expr/src/functions.rs:2879]: 👍 

BTW this is basically what I need to do in IOx / why I made https://github.com/apache/arrow-datafusion/issues/3708

**Review by alamb — COMMENTED** [review]
Looks good to me -- thanks @liukun4515

**Pull request merged**